### ISSUE TITLE
d3-request Definitions

### ISSUE THREAD

**Issue by tomwanzek** [tomwanzek/d3-v4-definitelytyped#44]
Add Definitions for **d3-request** module.

Important: Under this milestone _Data Acquisition (Enhanced)_. The definitions should be structured as proper modules/UMD-style module definitions. I.e., they should be comparable to the modules tracked under milestone _Core Modules (Enhanced)_ and contained in the  master branch of this repo.

Combined with their tests, they should compile successfully using Typescript 2.0.0 (beta).

They should make use of function-context `this` typing, where appropriate, comparable to its use in under the _Core Modules (Enhanced)_ milestone.

For future consideration, `null` and `undefined` types should be used, where the API indicates the explicit types to be, e.g. for a function return type, `string | null` vs `string | undefined` vs. `string` only for non-nullable defined types.

Note, however, that, currently and in correspondence to the approach taken by **DefinitelyTyped/types-2.0** the compiler is using the `"strictNullChecks": false` option.

**Comment by tomwanzek** [comment]
Added definitions from PR #52 by @Ledragon.

I have changed the naming convention for the generics to `RequestData` and `ResponseData`  from the original PR. This is to avoid confusion with other uses of the concept of _request type_ and i.p. _reponse type_ in the module definition.

Also, we need to review the `DSVRequest` interface and check the parameterization of the `row(...)` method. `RequestData` has no part of it. This should be something like `<ResponseDatum, MappedResponseDatum>`. We need to look at API/code.

I am going to uncheck the test box in README for now to indicate that the module definition is not 100% complete.

As an aside, so far we have left the `Request` interface free of a generic parameterization, while this makes it easier to handle flexibility at the method level. It does not enforce consistency across methods.

**Comment by Ledragon** [comment]
Indeed `DsvRequest` should not include any `RequestData` as these have nothing to do there. I like your idea of `<ResponseDatum, MappedResponseDatum>`; I will let it sink in.

Indeed the request interface could be made generic, something I did not think about upfront, but as all generating methods have a strongly typed argument, that would make total sense.

**Comment by tomwanzek** [comment]
The reason I did not consider making `Request<...>` generic right away, was simply that I am not that familiar with how folks are using it. As I said, I previously mostly used Angular's `$http` or `$resource` for the data side of things... (the `node` side of things completely aside...)

Given that one can use `d3-request` for different actions (GET, PUT, ...) The data submitted with the request, if any, and the data received in the response, if any, would differ anyway. So templating the overall interface may not be to sensible. And for now it may be easier to just deal with it as consistent naming of generics across the different methods, without cross-method enforcement?

**Comment by tomwanzek** [comment]
FYI, I have consolidated some milestones for issue tracking/prioritization. All _Standard Bundle Modules_ for D3 have been combined into a single milestone per TypeScript Version of the definition/tests. As a result, I removed a few more granular milestones primarily meant for my own work load prioritization.

**Comment by Ledragon** [comment]
Maybe having a type on the interface for the `ResponseData`, but not `RequestData`, the same way `angular.IPromise<T>` is returned when using `$http`?

**Comment by tomwanzek** [comment]
I looked into it and created a new branch `d3-request-finalization`. I made some initial suggested changes to the definitions see [here](https://github.com/tomwanzek/d3-v4-definitelytyped/blob/d3-request-finalization/src/d3-request/index.d.ts).

The issue that I see at this point is that adding generics to the interfaces `Request` and `DSVRequest` seems to add complication with predicting, how people want to use it in practice. I.e. there is no guarantee that people would use a request object comparable to a resource for a REST API. So imposing too much, may be making too many assumptions. Your comparison to Angular is somewhat instructive (talking about Angular 1.x here with `$http`, but also with the more elaborate `$resource`).

The major suggested fixes I made so far are as follows:
- I gave the `DSVRequest` related signatures a makeover, now that d3-dsv is finished. Since d3-request formally has a dependency to d3-dsv, we can readily import it. (This means some of the initial tests fail right now, as they do not adhere to the signature) I checked the source code, it invokes the `csvParse` and `tsvParse`, respectively.
- I went through the source code to look into the signatures for the event listeners and the context which is bound to them and the callbacks. I broke them out based on that understanding and added `this` typing as well. I know the D3 API documentation is not very explicit in this regard. This might be a worthwhile follow-up with Mike Bostock. Additional detail might be helpful and then there is the misguiding example in the doc which seems to imply that `load` listeners are called with two arguments (i.e. node-style)

Instead of the generics at the interface level, it might simply suffice to add and beef up JSDoc comments. This is most critical for the interaction of the callbacks and `load`-event listener in relation to setting a value accessor with `response(...)` and/or adding a `row`-mapper for the `DSVRequests`. For `csv` and `dsv`, the convenience DSV request factories _with_ callbacks are internally consistent now. The problem arises, when the `url`-only signature is used and the other configuration/get-steps are performed separately on the returned object.

Anyway, will need to update the tests to reflect the current thinking.
